File: 148.sort-list.py
#
# @lc app=leetcode id=148 lang=python3
#
# [148] Sort List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        if head == None or head.next == None:
            return head

        fast = head
        slow = head
        temp = head

        while fast != None and fast.next != None:
            temp = slow
            fast = fast.next.next
            slow = slow.next

        temp.next = None

        left = self(head)
        right = self(slow)

        def merge(l1: Optional[ListNode], l2: Optional[listNode]):
            temp = listNode(0)
            cur = temp

            while l1 != None and l2 != None:
                if l1.val < l2.val:
                    cur.next = l1
                    l1 = l1.next
                else:
                    cur.next = l2
                    l2 = l2.next

                cur = cur.next

            if l1 != None:
                cur.next = l1
            if l2 != None:
                cur.next = l2

            return temp

        return merge(left, right)


# @lc code=end

File: 148.sort-list.java
/*
 * @lc app=leetcode id=148 lang=java
 *
 * [148] Sort List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
public ListNode sortList(ListNode head) {
    return mergeSort(head);
}

public ListNode mergeSort(ListNode head){
    if(head == null || head.next == null){
        return head;
    }
    ListNode l = head;
    ListNode r = head;
    ListNode temp = head;
    while(r != null && r.next != null){
        temp = l;
        l = l.next;
        r = r.next.next;
    }
    temp.next = null;
    ListNode leftP = mergeSort(head);
    ListNode rightP = mergeSort(l);
    return merge(leftP, rightP);
}

public ListNode merge(ListNode l, ListNode r){
    ListNode dummy = new ListNode(-1);
    ListNode current = dummy;
    while(l != null && r != null){
        if(l.val > r.val){
            current.next = r;
            r = r.next;
        }else{
            current.next = l;
            l = l.next;
        }
        current = current.next;
    }

    if(l != null){
        current.next = l;
    }
    if(r != null){
        current.next = r;
    }
    return dummy.next;
}
// @lc code=end
